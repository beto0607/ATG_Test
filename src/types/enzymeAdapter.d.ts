// To avoid a typescript error with enzyme-adaoter-react-16
declare module 'enzyme-adapter-react-16';