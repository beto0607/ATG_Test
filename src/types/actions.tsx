import { GameSchedule, GameData } from "../types";

export const GAME_SCHEDULE_LOADED = "GAME_SCHEDULE_LOADED";
export type GAME_SCHEDULE_LOADED = typeof GAME_SCHEDULE_LOADED;

export const GAME_DATA_LOADED = "GAME_DATA_LOADED";
export type GAME_DATA_LOADED = typeof GAME_DATA_LOADED;


export interface GameScheduleLoadedAction {
    type: typeof GAME_SCHEDULE_LOADED;
    text: string;
    gameSchedule?: GameScheduleÂ | null;
}

export interface GameDataLoadedAction {
    type: typeof GAME_DATA_LOADED;
    gameData?: GameData;
}

export type LoadedActions = GameScheduleLoadedAction | GameDataLoadedAction;

export type AppActions = LoadedActions;