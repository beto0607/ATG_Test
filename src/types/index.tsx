type TrackInfo = {
    id: number,
    name: string
};
type FavoriteInfo = {
    raceId: string,
    legNumber: number,
    start: {
        number: number,
        name: string,
        distribution: number
    }
};
type GameInfo = {
    id: string,
    startTime: string, // YYYY-MM-DDThh:mm:ss
    tracks: Array<TrackInfo>,
    favorites: Array<FavoriteInfo>
}
//ENDPOINT RESPONSE
//https://www.atg.se/services/racinginfo/v1/api/products/<gameType>
type GameSchedule = {
    betType: string,
    upcoming: Array<GameInfo>,
    results: Array<GameInfo>
};

type RaceStart = {
    number: number,
    driver: {
        firstName: string,
        lastName: string
    },
    horse: {
        name: string,
        trainer: {
            firstName: string,
            lastName: string
        },
        pedigree: {
            father: {
                name: string
            }
        }
    }
}
type GameRace = {
    date: string,
    name: string,
    scheduledStartTime: string,
    starts: Array<RaceStart>
}
//ENDPOINT RESPONSE
// https://www.atg.se/services/racinginfo/v1/api/games/<gameId>
type GameResponse = {
    id: string,
    status: string,
    races: Array<GameRace>
};

export interface ApplicationState {
    text: string;
    game_schedule: GameSchedule;
    game_data: GameResponse;
}