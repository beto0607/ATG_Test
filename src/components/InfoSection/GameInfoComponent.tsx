import React, { useState } from "react";
import { GameInfo } from '../../types/index';

const GameInfoComponent: React.FC<GameInfo> = ({ id, startTime, tracks, favorites }: GameInfo) => {
    return (
        <li>
            <span>#{id}</span>
        </li>
    );
};

export default GameInfoComponent;
