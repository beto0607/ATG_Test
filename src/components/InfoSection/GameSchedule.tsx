import React from "react";
import { GameInfo } from '../../types/index';
import { ListGroup } from "react-bootstrap";
import GameInfoComponent from './GameInfoComponent';
import Moment from 'react-moment';
import styles from './info-section.module.scss';

export interface GameScheduleProps {
    title: string;
    games: Array<GameInfo>;
}
export interface GameInfoDayProps {
    day: string;
    games: Array<GameInfo>;
}

// Get the first part of a ISO Date type string
export function getDateFromString(text: string): string { return text.split('T')[0] }

// Splits the games array into several arrays by day(YYYY-MM-DD)
// The games of the same day will be in the same array
export function groupByDay(games: Array<GameInfo>): Array<GameInfoDayProps> {
    let returnValue: Array<GameInfoDayProps> = [];
    // Get the dates without duplicates
    let dates: Set<string> = new Set(games.map((element) => getDateFromString(element.startTime)));
    // Create the array
    dates.forEach((value: string) => {
        returnValue.push({
            day: value,
            // Gets the games for a day 
            games: games.filter(e => getDateFromString(e.startTime) === value)
        });
    })
    return returnValue;
}
export const GameInfoDay: React.FC<GameInfoDayProps> = ({ day, games }: GameInfoDayProps) => {
    return (
        <div className={styles['game-info-day-container']}>
            <Moment format="DD-MM-YYYY" date={day} className={styles['day']} />
            <ListGroup className={styles['list-group']}>
                {games.map((element) => <GameInfoComponent key={element.id} {...element} />)}
            </ListGroup>
        </div>
    );
};
export const GameSchedule: React.FC<GameScheduleProps> = ({ title, games }: GameScheduleProps) => {
    const gameByDate: Array<GameInfoDayProps> = groupByDay(games);
    return (
        <div className={styles['']}>
            <h3>
                {title}
            </h3>
            {gameByDate.map((value: GameInfoDayProps, index: number) => <GameInfoDay key={index} {...value} />)}
        </div>
    );
};

export default GameSchedule;
