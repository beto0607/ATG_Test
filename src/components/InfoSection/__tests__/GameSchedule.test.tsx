import React from 'react';
import { shallow } from 'enzyme';
import GameSchedule, { getDateFromString, groupByDay, GameInfoDayProps, GameInfoDay, GameScheduleProps } from '../GameSchedule';
import toJson from 'enzyme-to-json'
import { GameInfo } from '../../../types/index';
import { } from 'react-bootstrap';
import GameInfoComponent from '../GameInfoComponent';

const dateExample: string = "2019-07-24T11:20:04.492Z";
const dateExample2: string = "2019-07-25T11:20:04.492Z";
const mockGameInfo: GameInfo = {
    id: "1",
    startTime: dateExample,
    tracks: [],
    favorites: []
};
const mockGameInfoDayProps: GameInfoDayProps = {
    day: '2019-07-24',
    games: []
};
const mockGameScheduleProps: GameScheduleProps = {
    title: 'a title',
    games: []
};
describe('getDateFromString ', () => {
    it('should return the first part of an ISO date string', () => {
        expect(getDateFromString(dateExample)).toEqual('2019-07-24');
    });
});
describe('groupByDay', () => {
    it('should return an array with length 0', () => {
        const games: Array<GameInfo> = [];
        expect(groupByDay(games)).toHaveLength(0);
    });
    it('should return an array with length 2', () => {
        const games: Array<GameInfo> = [
            {
                ...mockGameInfo,
            },
            {
                ...mockGameInfo,
                id: "2",
                startTime: dateExample2
            }
        ];
        expect(groupByDay(games)).toHaveLength(2);
    });
});
describe('GameInfoDay', () => {
    it('renders without crashing', () => {
        const props: GameInfoDayProps = mockGameInfoDayProps;
        const component = shallow((<GameInfoDay {...props} />));
        expect(toJson(component)).toMatchSnapshot()
    });
    it('renders one GamInfoComponent', () => {
        const props: GameInfoDayProps = {
            ...mockGameInfoDayProps,
            games: [mockGameInfo]
        };
        const component = shallow((<GameInfoDay {...props} />));
        expect(component.find(GameInfoComponent)).toHaveLength(1)
    });
});
describe('GameSchedule', () => {
    it('renders without chashing', () => {
        const props: GameScheduleProps = mockGameScheduleProps;
        const component = shallow((<GameSchedule {...props} />));
        expect(toJson(component)).toMatchSnapshot()
    });
    it('should render 2 GameIfoDay components', () => {
        const games: Array<GameInfo> = [
            {
                ...mockGameInfo,
            },
            {
                ...mockGameInfo,
                id: "2",
                startTime: dateExample2
            }
        ];
        const props: GameScheduleProps = {
            ...mockGameScheduleProps,
            games: games
        };
        const component = shallow((<GameSchedule {...props} />));
        expect(component.find(GameInfoDay)).toHaveLength(2);
    });
});