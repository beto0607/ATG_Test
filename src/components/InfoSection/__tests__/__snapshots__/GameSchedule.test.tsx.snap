// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameInfoDay renders without crashing 1`] = `
<div
  className="game-info-day-container"
>
  <t
    ago={false}
    calendar={false}
    className="day"
    date="2019-07-24"
    decimal={false}
    element={null}
    filter={[Function]}
    format="DD-MM-YYYY"
    fromNow={false}
    interval={60000}
    local={false}
    onChange={[Function]}
    titleFormat=""
    toNow={false}
    unit={null}
    unix={false}
    utc={false}
    withTitle={false}
  />
  <ForwardRef
    className="list-group"
  />
</div>
`;

exports[`GameSchedule renders without chashing 1`] = `
<div
  className=""
>
  <h3>
    a title
  </h3>
</div>
`;
